### PR TITLE
Enhance maximum number of supported interfaces from 2^16.

### PR DESCRIPTION
**Pull request by tastywheattasteslikechicken**
Fixes https://bugzilla.xamarin.com/show_bug.cgi?id=10784

### CODE REVIEW

File: mono/metadata/class.c [hunk 1]
```diff
@@ -2781,7 +2781,7 @@ mono_unload_interface_id (MonoClass *klass)
  * LOCKING: Acquires the classes lock.
  * Returns: the new ID.
  */
-static guint
+static guint32
 mono_get_unique_iid (MonoClass *klass)
 {
 	int iid;
```
Comment (Therzok) [inline mono/metadata/class.c]: Shouldn't this be changed to `guint32`?

File: mono/metadata/class.c [hunk 2]
```diff
@@ -2834,7 +2834,9 @@ mono_get_unique_iid (MonoClass *klass)
 	}
 #endif
 
-	g_assert (iid <= 65535);
+	/* I've confirmed iids safe past 16 bits, however bitset code uses a signed int while testing.
+	 * Once this changes, it should be safe for us to allow 2^32-1 interfaces, until then 2^31-2 is the max. */
+	g_assert (iid < INT_MAX);
```
Comment (Therzok) [inline mono/metadata/class.c]: Coupled with the above change, shouldn't this be G_MAXUINT32?

**Comment by Therzok** [comment]
Added some comments, but would like someone from the runtime team to confirm.

File: mono/metadata/class.c (hunk 2 above)
Comment (tastywheattasteslikechicken) [inline mono/metadata/class.c]: Bitset code is signed, so I left that signed rather than unsigned,

File: mono/metadata/class.c (hunk 1 above)
Comment (tastywheattasteslikechicken) [inline mono/metadata/class.c]: What do you mean?

**Comment by evincarofautumn** [comment]
builddeb

**Comment by akoeplinger** [comment]
builddeb

**Comment by lewurm** [comment]
`<botmode>`
Benchmark results: http://xamarin.github.io/benchmarker/front-end/pullrequest.html#id=39
buildbot logs: https://performancebot.mono-project.com/builders/debian-amd64_benchmarker_pullrequest_auto-sgen-conc-noturbo-binary/builds/7
`</botmode>`

**Comment by lewurm** [comment]
`<botmode>`
Benchmark results: http://xamarin.github.io/benchmarker/front-end/pullrequest.html#id=40
buildbot logs: https://performancebot.mono-project.com/builders/debian-amd64_benchmarker_pullrequest_auto-sgen-noturbo-binary/builds/8
`</botmode>`

**Comment by lewurm** [comment]
`<botmode>`
Benchmark results: http://xamarin.github.io/benchmarker/front-end/pullrequest.html#id=41
buildbot logs: https://performancebot.mono-project.com/builders/debian-amd64_benchmarker_pullrequest_auto-sgen-conc-noturbo-binary/builds/8
`</botmode>`

**Comment by lewurm** [comment]
`<botmode>`
Benchmark results: http://xamarin.github.io/benchmarker/front-end/pullrequest.html#id=42
buildbot logs: https://performancebot.mono-project.com/builders/debian-amd64_benchmarker_pullrequest_auto-sgen-noturbo-binary/builds/9
`</botmode>`

**Comment by lewurm** [comment]
`<botmode>`
Benchmark results: http://xamarin.github.io/benchmarker/front-end/pullrequest.html#id=43
buildbot logs: https://performancebot.mono-project.com/builders/debian-amd64_benchmarker_pullrequest_auto-sgen-conc-noturbo-binary/builds/9
`</botmode>`

**Comment by lewurm** [comment]
`<botmode>`
Benchmark results: http://xamarin.github.io/benchmarker/front-end/pullrequest.html#id=44
buildbot logs: https://performancebot.mono-project.com/builders/debian-amd64_benchmarker_pullrequest_auto-sgen-noturbo-binary/builds/10
`</botmode>`

**Comment by tastywheattasteslikechicken** [comment]
Just checking up, is there anything more that needs to be done here? (I haven't heard back about the CLA I submitted yet)

**Comment by akoeplinger** [comment]
@tastywheattasteslikechicken the CLA wasn't signed by your manager (Brad L.) yet.

As for the PR itself: we discussed this internally and would like to get it in, but we need to do a proper review/audit first.

**Comment by tastywheattasteslikechicken** [comment]
We're waiting on legal.

**Comment by lewurm** [comment]
@tastywheattasteslikechicken any updates on your CLA?

**Comment by tastywheattasteslikechicken** [comment]
Believe it or not, I'm still waiting. About two weeks ago I was helpfully told legal are on it. What they were up to the other weeks I have no clue. I'll chase it up again today.

**Comment by tastywheattasteslikechicken** [comment]
Is it sufficient to get them to sign something releasing it under the MIT license?

**Comment by akoeplinger** [comment]
We're now under the .NET Foundation, so I guess you could also sign that one too (which afaik is a bit simpler): https://cla2.dotnetfoundation.org/

**Comment by migueldeicaza** [comment]
This probably needs to update the AOT image version.

**Comment by tastywheattasteslikechicken** [comment]
Our legal department has now (2 days ago) agreed to sign the CLA. We soon will have it available to send.

**Comment by shalupov** [comment]
Any news on this PR? We need it in our production too.

**Comment by tastywheattasteslikechicken** [comment]
The .NET Foundation one was fully online, and I've completed it (if I did it right)

**Comment by akoeplinger** [comment]
approve

**Comment by akoeplinger** [comment]
Hmm, looks like the  .NET Foundation CLA bot didn't recognize it. I'll talk to the owners on Monday about resolving that.

**Comment by osonmez** [comment]
Hello All,
Is this fix released?

**Comment by akoeplinger** [comment]
@osonmez it didn't make it into the upcoming Mono 4.8 release but it will be included in the one after that.

**Comment by osonmez** [comment]
@akoeplinger thanks Alexander

**Pull request merged**